docs/dev/topology-over-raft.md

```diff
@@ -224,3 +224,64 @@ There are also a few static columns for cluster-global properties:
 - `unpublished_cdc_generations` - the IDs of the committed yet unpublished CDC generations
 - `global_topology_request` - if set, contains one of the supported global topology requests
 - `new_cdc_generation_data_uuid` - used in `commit_cdc_generation` state, the time UUID of the generation to be committed
+
+# Join procedure
+
+In topology on raft mode, new nodes need to go through a new handshake procedure
+before they can join the cluster, including joining group 0. The handshake
+happens during raft discovery and replaced the old `GROUP0_MODIFY_CONFIG`.
+
+Two RPC verbs are introduced:
+
+- `JOIN_NODE_REQUEST`
+  Sent by the node that wishes to join the cluster. It contains some basic
+  information about the node that will have to be verified. Some of them
+  can be verified by the node that receives the RPC - which can result in
+  an immediate failure of the RPC -  but others need to be verified
+  by the topology coordinator.
+
+- `JOIN_NODE_REPLY`
+  Sent by the topology coordinator back to the joining node. It contains
+  coordinator's decision about the node - whether the request was accepted
+  or rejected.
+
+The procedure is not retryable. If the joining node crashes before finishing it,
+it might get rejected after restart. In order to retry adding the node, the data
+directory must be deleted first.
+
+*The procedure*
+
+If the node didn't join group 0, it sends `JOIN_NODE_REQUEST` to any existing
+node in the cluster. The receiving node can either:
+- Accept the request and tell the new node to wait for `JOIN_NODE_RESPONSE`.
+- Reject the request. This can happen if:
+  - The request does not satisfy some validity checks done by the receiving node
+    (e.g. cluster name does not match),
+  - There already an existing request to join this node (which can happen if
+    the node crashed during the previous attempt to join),
+  - The node is already a part of the cluster,
+  - The node was removed from the cluster.
+
+The new node should not process `JOIN_NODE_RESPONSE` RPC until it is explicitly
+ordered by node that handles `JOIN_NODE_REQUEST`. This is necessary so that
+it doesn't accidentally process a join request from a previous attempt, if there
+was any. Similarly, the group 0 server must not be started on the new node
+until the `JOIN_NODE_REQUEST` RPC succeeds, otherwise it might receive some
```

Comment:
`JOIN_NODE_REQUEST` or `JOIN_NODE_RESPONSE`? Only after coordinator decided that it is safe can we start group0, no?

Reply:
> `JOIN_NODE_REQUEST` or `JOIN_NODE_RESPONSE`?

`JOIN_NODE_REQUEST`.

> Only after coordinator decided that it is safe can we start group0, no?

The coordinator will either:

- Decide to accept us, add us to the configuration and will send `JOIN_NODE_RESPONSE` to us that we are accepted,
- Decide to reject us, _not_ add us to the configuration and will send `JOIN_NODE_RESPONSE` to us that we are rejected.

Therefore we will only receive group0 commands if we get accepted.

I suppose we could start group0 server when we receive `JOIN_NODE_RESPONSE`, but we are added to group0 configuration by the coordinator before it sends `JOIN_NODE_RESPONSE`, so some commands might be rejected because we didn't start group0 server yet.

Reply:
Command will be retired. The logic above makes sense though. I wanted just make sure the the doc is accurate. You can even extend it with the explanation above why it is safe.

But according to the same logic it is safe to create group0 server even before `JOIN_NODE_REQUEST` succeeds, no?

Reply:
> But according to the same logic it is safe to create group0 server even before JOIN_NODE_REQUEST succeeds, no?

I had the same question, we discussed this with Piotr D on slack :D

IIUC, this server might have been added to group 0 configuration on the previous join attempt. Now, if we started the server before JOIN_NODE_REQUEST, we'd start receiving commands, which would be incorrect (we might have restarted with a smaller set of features so we don't understand some commands). But if we start only after successful JOIN_NODE_REQUEST, in the case of second attempt, we would get rejected and never start the server, so we're protected from commands that we may not understand.